seatunnel-connectors-v2/connector-cdc/connector-cdc-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerStreamingChangeEventSource.java

```diff
@@ -110,265 +119,285 @@ public LogMinerStreamingChangeEventSource(
+     * time the snapshot was taken, we'd miss the events in those transactions that have an SCN
+     * smaller than the snapshot SCN.
+     *
+     * @param offsetContext the offset context
+     * @param firstScn the oldest SCN still available in the REDO logs
+     */
+    private void computeStartScnForFirstMiningSession(
+            OracleOffsetContext offsetContext, Scn firstScn) {
+        // This is the initial run of the streaming change event source.
+        // We need to compute the correct start offset for mining. That is not the snapshot offset,
+        // but the start offset of the oldest transaction that was still pending when the snapshot
+        // was taken.
+        Map<String, Scn> snapshotPendingTransactions =
+                offsetContext.getSnapshotPendingTransactions();
+        if (snapshotPendingTransactions == null || snapshotPendingTransactions.isEmpty()) {
+            // no pending transactions, we can start mining from the snapshot SCN
+            startScn = snapshotScn;
+        } else {
+            // find the oldest transaction we can still fully process, and start from there.
+            Scn minScn = snapshotScn;
+            for (Map.Entry<String, Scn> entry : snapshotPendingTransactions.entrySet()) {
+                String transactionId = entry.getKey();
+                Scn scn = entry.getValue();
+                LOGGER.info(
+                        "Transaction {} was pending across snapshot boundary. Start SCN = {}, snapshot SCN = {}",
+                        transactionId,
+                        scn,
+                        startScn);
+                if (scn.compareTo(firstScn) < 0) {
+                    LOGGER.warn(
+                            "Transaction {} was still ongoing while snapshot was taken, but is no longer completely recorded in the archive logs. Events will be lost. Oldest SCN in logs = {}, TX start SCN = {}",
+                            transactionId,
+                            firstScn,
+                            scn);
+                    minScn = firstScn;
+                } else if (scn.compareTo(minScn) < 0) {
+                    minScn = scn;
+                }
+            }
+
+            // Make sure the commit SCN is at least the snapshot SCN - 1.
+            // This ensures we'll never emit events for transactions that were complete before the
+            // snapshot was
+            // taken.
+            if (offsetContext.getCommitScn().compareTo(snapshotScn) < 0) {
+                LOGGER.info(
+                        "Setting commit SCN to {} (snapshot SCN - 1) to ensure we don't double-emit events from pre-snapshot transactions.",
+                        snapshotScn.subtract(Scn.ONE));
+                offsetContext
+                        .getCommitScn()
+                        .setCommitScnOnAllThreads(snapshotScn.subtract(Scn.ONE));
+            }
+
+            // set start SCN to minScn
+            if (minScn.compareTo(startScn) < 0) {
```

Comment:
I'd like to inquire about a question. I'm a bit confused about the conditions here. It seems like Flink CDC and Debezium might handle them differently.

```java
# debezium 1.9.8.Final

            // set start SCN to minScn
            if (minScn.compareTo(startScn) < 0) {
                LOGGER.info("Resetting start SCN from {} (snapshot SCN) to {} (start of oldest complete pending transaction)", startScn, minScn);
                startScn = minScn.subtract(Scn.ONE);
            }
```

```java
# flink-cdc

            // set start SCN to minScn
            if (minScn.compareTo(startScn) <= 0) {
                LOGGER.info(
                        "Resetting start SCN from {} (snapshot SCN) to {} (start of oldest complete pending transaction)",
                        startScn,
                        minScn);
                startScn = minScn.subtract(Scn.ONE);
            }
```

Reply:
Other than that, I don't think it makes any difference whether it's less than or equal to. Because the code logic is set start SCN to minScn. If their values are already the same, an assignment is no different from no assignment.

Reply:
Thanks